### PR TITLE
Add gitlab-CI

### PR DESCRIPTION
**Pull request by kerel-fs**
This PR is depends on #240.

todo:
- [x] Fix KOBO target (build log: https://gitlab.com/kerel-fs/xcsoar/-/jobs/271564405 )
- [x] Add upload of build artifacts to download.xcsoar.org

edit:
20190815: fixed ToDos.

### CODE REVIEW

**Comment by kerel-fs** [comment]
All CI builds finish successfully, see https://gitlab.com/kerel-fs/xcsoar/pipelines/76549132

The only untested part of these patches is the deployment of nightlies using gitlab-CI to download.xcsoar.org. This will require a gitlab repository which mirrors xcsoar/xcsoar and correctly configured CI variables.

**Comment by MaxKellermann** [comment]
This PR contains unrelated commits from your other PR

**Comment by kerel-fs** [comment]
Correct. Fortunately you merged the branch of my other PR using fast-forward, so it's not a problem (otherwise I would've rebased this branch).

File: ide/provisioning/install-debian-packages.sh [hunk 1]
```diff
@@ -2,8 +2,12 @@
 
 set -e
 
+# To prevent apt from displaying dialogs
+DEBIAN_FRONTEND=noninteractive
```
Comment (MaxKellermann) [inline ide/provisioning/install-debian-packages.sh]: Unrelated change, and one without effect.

File: ide/provisioning/install-debian-packages.sh [hunk 2]
```diff
@@ -2,8 +2,12 @@
...
+APTOPTS="--no-install-recommends --assume-yes"
+
 echo Installing base dependencies...
-apt-get --assume-yes install make \
+apt-get install $APTOPTS make \
```
Comment (MaxKellermann) [inline ide/provisioning/install-debian-packages.sh]: Your commit not only adds `--no-install-recommends`, but also refactors the existing `--assume-yes` into a global variable. This makes the commit obscure.

File: ide/provisioning/install-debian-packages.sh [hunk 3]
```diff
@@ -13,9 +13,10 @@ echo Installing base dependencies...
 apt-get install $APTOPTS make \
   librsvg2-bin xsltproc \
   imagemagick gettext ffmpeg \
-  git quilt zip \
+  git quilt unzip zip \
```
Comment (MaxKellermann) [inline ide/provisioning/install-debian-packages.sh]: What for?

File: ide/provisioning/install-debian-packages.sh [hunk 4]
```diff
@@ -13,9 +13,10 @@ echo Installing base dependencies...
...
   m4 automake wget \
-  ttf-bitstream-vera fakeroot
+  ttf-bitstream-vera fakeroot \
+  ccache
```
Comment (MaxKellermann) [inline ide/provisioning/install-debian-packages.sh]: What for?

File: ide/provisioning/install-debian-packages.sh [hunk 5]
```diff
@@ -42,11 +43,12 @@ apt-get install $APTOPTS make g++ \
   librsvg2-bin xsltproc \
   imagemagick gettext \
   mesa-common-dev libgl1-mesa-dev libegl1-mesa-dev \
-  fonts-dejavu
+  fonts-dejavu \
+  libnetcdf-dev
```
Comment (MaxKellermann) [inline ide/provisioning/install-debian-packages.sh]: What for?

File: ide/provisioning/install-debian-packages.sh [hunk 6]
```diff
@@ -42,11 +43,12 @@ apt-get install $APTOPTS make g++ \
...
 echo
 
 echo Installing dependencies for compiling with LLVM / Clang...
-apt-get install $APTOPTS llvm clang libc++-dev
+apt-get install $APTOPTS llvm clang-4.0 libc++-dev
```
Comment (MaxKellermann) [inline ide/provisioning/install-debian-packages.sh]: Why hard-code the clang version? This is fragile, and 4.0 is an ancient version.

**Comment by MaxKellermann** [comment]
Sigh, this PR adds yet another copy of dependency and build commands (to `ide/docker`). Note that I will not take care for keeping those up to date. If I ever push commits, I will not check if they break anything here. This is cumbersome and fragile.

File: ide/provisioning/install-debian-packages.sh [hunk 7]
```diff
@@ -44,11 +44,12 @@ apt-get install $APTOPTS make g++ \
...
   mesa-common-dev libgl1-mesa-dev libegl1-mesa-dev \
   fonts-dejavu \
-  libnetcdf-dev
+  libnetcdf-dev \
+  xz-utils
 echo
 
 echo Installing dependencies for compiling with LLVM / Clang...
-apt-get install $APTOPTS llvm clang-4.0 libc++-dev
+apt-get install $APTOPTS llvm clang libc++-dev
```
Comment (MaxKellermann) [inline ide/provisioning/install-debian-packages.sh]: Oh, now we're back to plain unversioned `clang`. Why submit a PR changing it forth and back?

File: ide/provisioning/install-debian-packages.sh [hunk 8]
```diff
@@ -64,7 +65,7 @@ apt-get install $APTOPTS g++-mingw-w64
 echo
 
 echo Installing dependencies for the Android target, not including SDK / NDK...
-apt-get install $APTOPTS default-jdk-headless vorbis-tools adb libtool
+apt-get install $APTOPTS openjdk-8-jdk-headless vorbis-tools adb libtool
```
Comment (MaxKellermann) [inline ide/provisioning/install-debian-packages.sh]: Hard-coding OpenJDK 8 will break building on Buster, which doesn't have 8.

File: ide/provisioning/install-debian-packages.sh [hunk 9]
```diff
@@ -68,10 +68,6 @@ echo Installing dependencies for the Android target, not including SDK / NDK...
 apt-get install $APTOPTS openjdk-8-jdk-headless vorbis-tools adb libtool
 echo
 
-echo Installing dependencies for KOBO
-apt-get --assume-yes install libav-tools
```
Comment (MaxKellermann) [inline ide/provisioning/install-debian-packages.sh]: This line was added in this very PR, and now remove it again? Why add it in the first place?

**Comment by MaxKellermann** [comment]
This PR doesn't make me happy. It is obscure, underdocumented, changes things forth and back, creeps in unrelated changes hidden in other commits.

**Comment by kerel-fs** [comment]
Thanks for your review!
I don't want to make you unhappy. :(
My intention is to automatically build as many targets as possible using a unmodified xcsoar source tree. I'll try to address all your annotations.

File: ide/provisioning/install-debian-packages.sh (hunk 1 above)
Comment (kerel-fs) [reply ide/provisioning/install-debian-packages.sh]: Fixed and moved into a separate commit.

File: ide/provisioning/install-debian-packages.sh (hunk 2 above)
Comment (kerel-fs) [reply ide/provisioning/install-debian-packages.sh]: Moved into a separate commit.

File: ide/provisioning/install-debian-packages.sh (hunk 4 above)
Comment (kerel-fs) [reply ide/provisioning/install-debian-packages.sh]: Removed, a left-over I didn't tidy up (introduced without explanation in https://github.com/lordfolken/xcsoar-docker-build/pull/1).

File: ide/provisioning/install-debian-packages.sh (hunk 5 above)
Comment (kerel-fs) [reply ide/provisioning/install-debian-packages.sh]: The upcoming Skysight integration will add this requirement, but as long as it's not merged we don't need it. Removed.

File: ide/provisioning/install-debian-packages.sh (hunk 6 above)
Comment (kerel-fs) [reply ide/provisioning/install-debian-packages.sh]: Sorry, yet another left-over from xcsoar-docker-build (using debian stretch). Removed.

File: ide/provisioning/install-debian-packages.sh (hunk 9 above)
Comment (kerel-fs) [reply ide/provisioning/install-debian-packages.sh]: > This commit incorporates all changes to install-debian-packages.sh
> available in the current master branch (bb41583) of
> xcsoar-docker-build [1].

As explained in the commit message c019eecc, I based these patches on lordfolken/xcsoar-docker-build@bb41583. Thus I first incorporated all changes in this repository and then added my patches.

As this clutters the xcsoar git history (and is harder to review, sry) I edited my patchset accordingly now.

File: ide/provisioning/install-debian-packages.sh (hunk 7 above)
Comment (kerel-fs) [reply ide/provisioning/install-debian-packages.sh]: See https://github.com/XCSoar/XCSoar/pull/241#discussion_r314697580. Removed.

File: ide/provisioning/install-debian-packages.sh (hunk 3 above)
Comment (kerel-fs) [reply ide/provisioning/install-debian-packages.sh]: `ide/provisioning/install-android-tools.sh` uses `unzip` to extract the Android NDK files. Moved into a separate commit.

**Comment by MaxKellermann** [comment]
You did not fix my complaint about the OpenJDK 8 switch;

> The currently used Android SDK Plaform Android 8.0 / API level 26 requires OpenJDK 8 [1].

Does it? I don't have OpenJDK 8, only 11, and I can build XCSoar. How come?

**Comment by kerel-fs** [comment]
> You did not fix my complaint about the OpenJDK 8 switch;

Correct, since I'm not finished updating this branch. :)

**Comment by kerel-fs** [comment]
> Does it? I don't have OpenJDK 8, only 11, and I can build XCSoar. How come?

Interesting, this is good news! I based my assumption on https://source.android.com/setup/build/older-versions#jdk which explicitly states that Android 8.0 (which is our target using android platform level 26) requires OpenJDK 8. I'll give OpenJDK another try (previously it failed).

**Comment by kerel-fs** [comment]
Using OpenJDK 11 the installation of the Android SDK using the sdkmanager fails.
https://github.com/XCSoar/XCSoar/blob/3cf15face691dc27985abaf555a16c2f099d7ed8/ide/provisioning/install-android-tools.sh#L36-L38

Error log:
```
Installing Android SDK packages...
Exception in thread "main" java.lang.NoClassDefFoundError: javax/xml/bind/annotation/XmlSchema
	at com.android.repository.api.SchemaModule$SchemaModuleVersion.<init>(SchemaModule.java:156)
	at com.android.repository.api.SchemaModule.<init>(SchemaModule.java:75)
	at com.android.sdklib.repository.AndroidSdkHandler.<clinit>(AndroidSdkHandler.java:81)
	at com.android.sdklib.tool.SdkManagerCli.main(SdkManagerCli.java:117)
	at com.android.sdklib.tool.SdkManagerCli.main(SdkManagerCli.java:93)
Caused by: java.lang.ClassNotFoundException: javax.xml.bind.annotation.XmlSchema
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 5 more
```

Afaik this is caused by using an incompatible OpenJDK version.
Maybe you can explain to me how to install the Android SDK package using OpenJDK 11?
There is an ugly [workaround](https://stackoverflow.com/a/55982976) available, but I expect there are better ways to do it.

If we do not find a fix or workaround for this, I'd like to recommend to install OpenJDK 8 from Debian unstable as done in this PR to provide setup scripts for a working build environment.

**Comment by kerel-fs** [comment]
>  Maybe you can explain to me how to install the Android SDK package using OpenJDK 11?

Answering the question myself:
The recommended way to install the Android SDK tools & platform is using the Android Studio (using a GUI).

Unfortunately for an unattended installation using the old `sdkmanager` CLI is still the recommended way, see [Update your tools with the SDK Manager
](https://developer.android.com/studio/intro/update#sdk-manager), so I guess we have to accept using OpenJDK 8 here.
